gnsi/acltrie/trie.go

```diff
@@ -0,0 +1,197 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package acltrie contains a trie for gNMI path with a corresponding policy.
+package acltrie
+
+import (
+	"errors"
+	"fmt"
+	"sort"
+	"strings"
+
+	gpb "github.com/openconfig/gnmi/proto/gnmi"
+	pathzpb "github.com/openconfig/gnsi/pathz"
+)
+
+// Trie is the root of the ACL trie.
+type Trie struct {
+	root *trieNode
+}
+
+type trieNode struct {
+	// childrenByRank orders the children of the node by number the set leaves for the element.
+	// This allows fast lookup of the best match: the one with the most number of set keys.
+	childrenByRank []map[string]*trieNode
+	elem           *gpb.PathElem
+	hasPolicy      bool
+	users          policies
+	groups         policies
+}
+
+// Insert inserts a new policy into the trie.
+func (t *Trie) Insert(r *pathzpb.AuthorizationRule) error {
+	if t.root == nil {
+		t.root = &trieNode{}
+	}
+	if r.Action == pathzpb.Action_ACTION_UNSPECIFIED {
+		return fmt.Errorf("action unspecified")
+	}
+	if r.Mode == pathzpb.Mode_MODE_UNSPECIFIED {
+		return fmt.Errorf("mode unspecified")
+	}
+	if r.GetGroup() == "" && r.GetUser() == "" {
+		return fmt.Errorf("principal unset")
+	}
+
+	path := r.GetPath()
+	node := t.root
+
+	for i, elem := range path.Elem {
+		if elem.Name == "*" {
+			return fmt.Errorf("wildcard path names are not permitted")
+		}
+
+		// Normalize path string by pruning wildcard keys.
+		pathStr, err := elemToString(elem.Name, elem.Key)
+		if err != nil {
+			return fmt.Errorf("invalid path element: %v", err)
+		}
+
+		// The rank of the child is the number of non-wildcards key (if any).
+		rank := len(setKeys(elem.Key))
+		if rank >= len(node.childrenByRank) {
+			node.childrenByRank = append(node.childrenByRank, make([]map[string]*trieNode, rank-len(node.childrenByRank)+1)...)
```

Comment:
I can't understand this part of the code. Can you explain this line?

Reply:
Suggest to make this a generic utility. e.g. s = slices.Grow(s, newLen-len(s))[:newLen]) -- example is inlined.

Reply:
slice.Grow only extends the capacity not the length of the slice

Reply:
The suggestion has `[:newLen]` which extends the length as well. Up to you, I think it improves readability.

Reply:
missed that in reading your comment, fixed